### PR TITLE
Related Posts: implement customization options in Customizer

### PR DESCRIPTION
**Pull request by eliorivero**
#### Changes proposed in this Pull Request:
- introduces the ability to customize the appearance of Related Posts in the Customizer
  captura de pantalla 2016-09-20 a las 16 24 38
- extends the options to show/hide date and/or context. The options and preview are added in Settings > Reading too.

relatedpostsreading
- when user is in the Customizer, dummy content is used to ensure that Related Posts are always visible to be customized
- Related Posts: remove settings in Jetpack React UI with link to Customizer focusing the Related Posts section, and loading the last published entry of 'post' type. After closing, user returns to Jetpack > Settings > Engagement.

jetpackreact

### CODE REVIEW

File: modules/related-posts/class.related-posts-customize.php [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	/**
+	 * Check that we're in a single post view.
+	 *
+	 * @since 4.2.0
+	 *
+	 * @return bool
+	 */
+	public static function is_single() {
```
Comment (ebinnion) [inline modules/related-posts/class.related-posts-customize.php]: Do we need this method? It doesn't seem to be used in this code. But, then again, I'm not super familiar with customizer.  ¯\_(ツ)_/¯

File: modules/related-posts/class.related-posts-customize.php [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	/**
+	 * Check that we're not in a single post view.
+	 *
+	 * @since 4.2.0
+	 *
+	 * @return bool
+	 */
+	public static function is_not_single() {
```
Comment (ebinnion) [inline modules/related-posts/class.related-posts-customize.php]: Same here, do we need this? Also, if we have the `is_single()` method, do we need a whole other method to negate that?

File: modules/related-posts/class.related-posts-customize.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+ * Control that displays a message in Customizer.
+ *
+ * @since 4.4
+ */
+class Jetpack_Message_Control extends WP_Customize_Control {
+
+	/**
+	 * Render the message.
+	 *
+	 * @since 4.4
```
Comment (ebinnion) [inline modules/related-posts/class.related-posts-customize.php]: This should be `4.4.0`.

File: modules/related-posts/class.related-posts-customize.php (hunk 1 above)
Comment (eliorivero) [inline modules/related-posts/class.related-posts-customize.php]: It's used in `'active_callback' => __CLASS__ . '::is_single',`

File: modules/related-posts/class.related-posts-customize.php (hunk 2 above)
Comment (eliorivero) [inline modules/related-posts/class.related-posts-customize.php]: It's necessary due to how it's passed to Customizer.

File: modules/related-posts/class.related-posts-customize.php [hunk 4]
```diff
@@ -0,0 +1,202 @@
+<?php
+/**
+ * Class to include elements to modify Related Posts look in Customizer.
+ *
+ * @since 4.2.0
```
Comment (eliorivero) [inline modules/related-posts/class.related-posts-customize.php]: All versions updated.

File: modules/related-posts/class.related-posts-customize.php [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+ */
+
```
Comment (eliorivero) [inline modules/related-posts/class.related-posts-customize.php]: Removed.

File: modules/related-posts/jetpack-related-posts.php [hunk 6]
```diff
@@ -316,6 +341,10 @@ public function print_setting_html() {
 			esc_html__( 'Show a "Related" header to more clearly separate the related section from posts', 'jetpack' ),
 			checked( $options['show_thumbnails'], true, false ),
 			esc_html__( 'Use a large and visually striking layout', 'jetpack' ),
+			checked( $options['show_date'], true, false ),
+			esc_html__( 'Show entry date', 'jetpack' ),
+			checked( $options['show_context'], true, false ),
+			esc_html__( "Show context (category or tag)", 'jetpack' ),
```
Comment (eliorivero) [inline modules/related-posts/jetpack-related-posts.php]: Changed.

File: modules/related-posts/related-posts.js [hunk 7]
```diff
@@ -173,30 +181,77 @@
...
+						startRelatedPosts();
+					}
+				} );
+			}
+			wp.customize.bind( 'preview-ready', startRelatedPosts );
+		} else {
+			startRelatedPosts();
+		}
 	} );
+
```
Comment (eliorivero) [inline modules/related-posts/related-posts.js]: Solved

**Comment by rcoll** [comment]
@eliorivero this could use a rebase, and I'll get it reviewed.

**Comment by samhotchkiss** [comment]
I'd love to get this in for 4.5-- @eliorivero can you please rebase ASAP and then ping some folks to test and merge quickly?

**Comment by eliorivero** [comment]
@rcoll it's rebased now, thanks for taking a look.

**Comment by ebinnion** [comment]
This PR needs a rebase.

**Comment by eliorivero** [comment]
Done, thanks for the heads up Eric 👍

File: _inc/client/components/module-settings/index.jsx [hunk 8]
```diff
@@ -68,69 +68,6 @@ export let SharedaddySettings = React.createClass( {
 
 SharedaddySettings = moduleSettingsForm( SharedaddySettings );
 
-export let RelatedPostsSettings = React.createClass( {
```
Comment (eliorivero) [inline _inc/client/components/module-settings/index.jsx]: @dereksmart I honestly don't think it's a regression. It's much better to tweak the options in the Customizer, seeing the related posts in context in the post and the site, being affected by any stylesheets the theme or other plugins might be loading than the sandboxed and inaccurate preview in the settings.
This is actually the whole point of this PR, if this doesn't sound good, let's close the PR.

Comment (dereksmart) [inline _inc/client/components/module-settings/index.jsx]: I'm sold :)

**Pull request merged**